tests/datamodules/test_experience_sources.py

```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
         self.agent = DummyAgent(net=self.net)
         self.env = gym.make("CartPole-v0")
         self.device = torch.device('cpu')
-        self.source = ExperienceSource(self.env, self.agent)
-
-    def test_step(self):
-        exp, reward, done = self.source.step(self.device)
-        self.assertEqual(len(exp), 5)
+        self.source = DummyExperienceSource(self.env, self.agent)
 
-    def test_episode(self):
-        total_reward = self.source.run_episode(self.device)
-        self.assertIsInstance(total_reward, float)
+    def test_dummy_base_class(self):
+        """Tests that base class is initialized correctly"""
+        self.assertTrue(isinstance(self.source.env, gym.Env))
+        self.assertTrue(isinstance(self.source.agent, Agent))
+        out = next(iter(self.source))
+        self.assertTrue(torch.all(out.eq(torch.ones(3))))
 
 
-class TestNStepExperienceSource(TestCase):
+class TestExperienceSource(TestCase):
 
     def setUp(self) -> None:
         self.net = Mock()
         self.agent = DummyAgent(net=self.net)
-        self.env = gym.make("CartPole-v0")
-        self.n_step = 2
-        self.source = NStepExperienceSource(self.env, self.agent, n_steps=self.n_step)
+        self.env = [gym.make("CartPole-v0") for _ in range(2)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent)
+
+    def test_init(self):
+        """Test that experience source is setup correctly"""
+        self.assertEqual(self.source.n_steps, 1)
+        self.assertIsInstance(self.source.pool, list)
+
+        self.assertEqual(len(self.source.states), len(self.source.pool))
+        self.assertEqual(len(self.source.histories), len(self.source.pool))
+        self.assertEqual(len(self.source.cur_rewards), len(self.source.pool))
+        self.assertEqual(len(self.source.cur_steps), len(self.source.pool))
+
+    def test_init_single_env(self):
+        """Test that if a single env is passed that it is wrapped in a list"""
+        single_env = Mock()
+        self.source = ExperienceSource(single_env, self.agent)
+        self.assertIsInstance(self.source.pool, list)
+
+    def test_env_actions(self):
+        """Assert that a list of actions of shape [num_envs, action_len] is returned"""
+        actions = self.source.env_actions(self.device)
+        self.assertEqual(len(actions), len(self.env))
+        self.assertTrue(isinstance(actions[0], list))
+
+    def test_env_step(self):
+        """Assert that taking a step through a single environment yields a list of history steps"""
+        actions = [[1], [1]]
+        env = self.env[0]
+        exp = self.source.env_step(0, env, actions[0])
+
+        self.assertTrue(isinstance(exp, Experience))
+
+    def test_source_next_single_env_single_step(self):
+        """Test that steps are executed correctly with one environment and 1 step"""
+
+        self.env = [gym.make("CartPole-v0") for _ in range(1)]
+        self.device = torch.device('cpu')
```

Comment:
do we need to have explicitly CPU or let's take the best available?

Reply:
No, it should take the best available, ill update it